### PR TITLE
Add indexOf to root.array

### PR DESCRIPTION
**Pull request by wilzbach**
For use in e.g. https://github.com/dlang/dmd/pull/8189#discussion_r182639975 or https://github.com/dlang/dmd/pull/8201/files#r183921790

### CODE REVIEW

File: src/dmd/root/array.d [hunk 1]
```diff
@@ -146,6 +146,35 @@ public:
...
+    /**
+    Search the array for an element and returns its index.
+    Params:
+        el = element to search for
+
+    Returns: index of the element, -1 otherwise
+    */
+    ptrdiff_t indexOf(T el) nothrow
+    {
+        foreach (i; 0 .. dim)
```
Comment (jacob-carlborg) [inline src/dmd/root/array.d]: Why not a regular `foreach` instead of a `foreach` range loop?

Comment (wilzbach) [reply src/dmd/root/array.d]: No reason. Thanks for the reminder!

**Review by jacob-carlborg — APPROVED** [review]

File: src/dmd/root/array.d [hunk 2]
```diff
@@ -146,6 +146,35 @@ public:
...
     }
 
+    /**
+    Search the array for an element and returns its index.
+    Params:
+        el = element to search for
+
+    Returns: index of the element, -1 otherwise
+    */
+    ptrdiff_t indexOf(T el) nothrow
```
Comment (WalterBright) [inline src/dmd/root/array.d]: `el` should be `const`, function should be `@nogc`

File: src/dmd/root/array.d [hunk 3]
```diff
@@ -146,6 +146,35 @@ public:
...
     }
 
+    /**
+    Search the array for an element and returns its index.
+    Params:
+        el = element to search for
+
+    Returns: index of the element, -1 otherwise
+    */
+    ptrdiff_t indexOf(const T el) nothrow @nogc
```
Comment (jacob-carlborg) [inline src/dmd/root/array.d]: `pure` as well?

**Comment by JinShil** [comment]
Grrrr!  DAutoTest uses a newer compiler version than auto-tester, and it throws the following error:

```
dmd/root/array.d(160): Error: none of the overloads of `opEquals` are callable using argument types `(Statement, const(Statement))`, candidates are:
```

I can't seem to produce this locally, either.

**Comment by jacob-carlborg** [comment]
Perhaps adding `const` to `v` in the `foreach` loop will help.

**Comment by jacob-carlborg** [comment]
Please squash commits when done.

**Comment by JinShil** [comment]
> dmd/root/array.d(160): Error: none of the overloads of `opEquals` are callable using argument types `(const(Statement), const(Statement))`, candidates are:
object.d(140):        `object.opEquals(Object lhs, Object rhs)`
object.d(165):        `object.opEquals(const(Object) lhs, const(Object) rhs)`

Great!

**Comment by wilzbach** [comment]
@JinShil looks like with the upgrade to 2.079, this is no longer blocked :)
Are you okay with squashing the commits into one?

**Comment by JinShil** [comment]
@wilzbach, It's actually your PR, so squash away.

**Comment by wilzbach** [comment]
> I wonder if it's a quirk of how DAutoTest chooses the version of Phobos to test with.

I think it's using `master`. I tried another repush.

> Would it help to update markdown-master to the tip of master, if it isn't already there?

This PR already tries to merge master into `markdown`.

CC @CyberShadow any other ideas what could have gone wrong here?

**Comment by CyberShadow** [comment]
Sorry, what do you mean? This PR is targeting branch master, not markdown.

**Comment by wilzbach** [comment]
Sorry this comment was supposed to be for https://github.com/dlang/dmd/pull/8325 - not sure how it ended up here :/

**Comment by wilzbach** [comment]
This is still blocked and I don't need it anymore :/

**Pull request closed without merging**